contracts/utils/HexUtils.sol

```
@@ -2,45 +2,90 @@
     function hexStringToBytes32(
-        bytes memory str,
-        uint256 idx,
-        uint256 lastIdx
-    ) internal pure returns (bytes32, bool) {
-        require(lastIdx - idx <= 64);
-        (bytes memory r, bool valid) = hexToBytes(str, idx, lastIdx);
-        if (!valid) {
-            return (bytes32(0), false);
+        bytes memory hexString,
+        uint256 pos,
+        uint256 end
+    ) internal pure returns (bytes32 word, bool valid) {
+        uint256 nibbles = end - pos;
+        if (nibbles > 64 || end > hexString.length) {
+            return (bytes32(0), false); // too large or out of bounds
         }
-        bytes32 ret;
+        uint256 src;
+        assembly {
+            src := add(add(hexString, 32), pos)
+        }
+        valid = unsafeBytes(src, 0, nibbles);
         assembly {
-            ret := shr(mul(4, sub(64, sub(lastIdx, idx))), mload(add(r, 32)))
+            let pad := sub(32, shr(1, add(nibbles, 1))) // number of bytes
+            word := shr(shl(3, pad), mload(0)) // right align
         }
-        return (ret, true);
     }
 
+    /// @dev Convert `hexString[pos:end]` to `address`.
+    ///      Accepts 40-64 hex-chars.
+    ///      Uses right alignment: `FF8000000000000000000000000000000000000001` &rarr; `8000000000000000000000000000000000000001`.
+    /// @param hexString The string to parse.
+    /// @param pos The index to start parsing.
+    /// @param end The (exclusive) index to stop parsing.
+    /// @return addr The parsed address.
+    /// @return valid True if the parse was successful.
+    function hexToAddress(
+        bytes memory hexString,
+        uint256 pos,
+        uint256 end
+    ) internal pure returns (address addr, bool valid) {
+        if (end - pos < 40) return (address(0), false); // too short
+        bytes32 word;
+        (word, valid) = hexStringToBytes32(hexString, pos, end);
+        addr = address(uint160(uint256(word)));
+    }
```

Does it make any sense to have this function? If it's >40 bytes, the high bytes will be silently discarded.

---

I kept the function behavior from before.  I figured it was just the companion to `addressToHex()`.

The 40 byte guard could be removed but this function does define the alignment, eg. `uint160` (right) not `bytes20` (left).

---

I think this function is unsafe as-is - it will only work safely for 40 byte addresses, so that's all we should support.

---

Just to be clear, the only change is odd-length support:
```                
                <-------------- 40 chars -------------->
hexToAddress(  "0000000000000000000000000000000000000001") = address(1) 
hexToAddress( "F0000000000000000000000000000000000000001") = address(1) // odd, failed before
hexToAddress("FF0000000000000000000000000000000000000001") = address(1)
```

So are you saying change to exactly 40 bytes? (instead of 40-64)

---

Good catch. Yes, I think we have to do that - silently dropping address bits is unsafe.